prow/cmd/branchprotector/protect.go

```diff
@@ -0,0 +1,293 @@
+			err := p.client.RemoveBranchProtection(r.Org, r.Repo, r.Branch)
+			if err != nil {
+				p.errors.add(fmt.Errorf("remove %s/%s=%s protection failed: %v", r.Org, r.Repo, r.Branch, err))
+			}
+		} else {
+			err := p.client.UpdateBranchProtection(r.Org, r.Repo, r.Branch, r.Contexts, r.Pushers)
+			if err != nil {
+				p.errors.add(fmt.Errorf("update %s/%s=%s protection failed: %v", r.Org, r.Repo, r.Branch, err))
+			}
+		}
+	}
+	p.done <- p.errors.errs
+}
+
+// Protect branches specified in the presubmit and branch-protection config sections.
+func (p *Protector) Protect() {
+	bp := p.cfg.BranchProtection
+
+	// Scan the branch-protection configuration
+	for orgName, org := range bp.Orgs {
+		if err := p.UpdateOrg(orgName, org, bp.Protect, bp.Contexts, bp.Pushers); err != nil {
+			p.errors.add(err)
+		}
+	}
+
+	// Do not automatically protect tested repositories
+	if !bp.ProtectTested {
+		return
+	}
+
+	// Some repos with presubmits might not be listed in the branch-protection
+	for repo := range p.cfg.Presubmits {
+		if p.completedRepos[repo] == true {
+			continue
+		}
+		parts := strings.Split(repo, "/")
+		if len(parts) != 2 {
+			log.Fatalf("Bad repo: %s", repo)
+		}
+		orgName := parts[0]
+		repoName := parts[1]
+		repoReqs := repoRequirements(orgName, repoName, *p.cfg)
+		protect := len(repoReqs) > 0
+		if err := p.UpdateRepo(orgName, repoName, config.Repo{}, &protect, repoReqs, nil); err != nil {
+			p.errors.add(err)
+		}
+	}
+}
+
+// Update all repos in the org with the specified defaults
+func (p *Protector) UpdateOrg(orgName string, org config.Org, protect *bool, contexts, pushers []string) error {
+	if org.Protect != nil {
+		protect = org.Protect
+	}
+
+	var oc []string
+	oc = append(oc, contexts...)
+	oc = append(oc, org.Contexts...)
+
+	var op []string
+	op = append(op, pushers...)
+	op = append(op, org.Pushers...)
+
+	repos, err := p.client.GetRepos(orgName, false)
+	if err != nil {
+		return fmt.Errorf("GetRepos(%s) failed: %v", orgName, err)
+	}
+	for _, repo := range repos {
+		repoName := repo.Name
+		err := p.UpdateRepo(orgName, repoName, org.Repos[orgName+"/"+repoName], protect, oc, op)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Update all branches in the repo with the specified defaults
+func (p *Protector) UpdateRepo(orgName string, repo string, repoDefaults config.Repo, protect *bool, contexts, pushers []string) error {
+	p.completedRepos[orgName+"/"+repo] = true
+
+	var rc []string
+	rc = append(rc, contexts...)
+	var rp []string
+	rp = append(rp, pushers...)
+
+	if repoDefaults.Protect != nil {
+		protect = repoDefaults.Protect
+	}
+	rc = append(rc, repoDefaults.Contexts...)
+	rp = append(rp, repoDefaults.Pushers...)
+
+	repoReqs := repoRequirements(orgName, repo, *p.cfg)
+	rc = append(rc, repoReqs...)
+
+	branches, err := p.client.GetBranches(orgName, repo)
+	if err != nil {
+		return fmt.Errorf("GetBranches(%s, %s) failed: %v", orgName, repo, err)
+	}
+	for _, branch := range branches {
```

Comment:
do we really want to apply these settings to *all* branches? sometimes admins click the revert button and create new ones, like [this](https://github.com/kubernetes/kubernetes/tree/revert-57401-automated-cherry-pick-of-%2357394-upstream-release-1.7).

Reply:
Note that I believe anyone with "write" can push a new branch.. doesn't have to be admin, I don't believe.

I'm not sure it necessitates a change now, but we should keep an eye on if the revert button causes problems. We already get Slack alerts when someone creates a new branch right on upstream, and nine times out of ten it's the revert button.